google-oauth-client/src/main/java/com/google/api/client/auth/openidconnect/IdTokenVerifier.java

```diff
@@ -280,5 +439,170 @@ public Builder setAcceptableTimeSkewSeconds(long acceptableTimeSkewSeconds) {
+
+    /**
+     * Data class used for deserializing a JSON Web Key Set (JWKS) from an external HTTP request.
+     */
+    public static class JsonWebKeySet extends GenericJson {
+      @Key public List<JsonWebKey> keys;
+    }
+
+    /** Data class used for deserializing a single JSON Web Key. */
+    public static class JsonWebKey {
+      @Key public String alg;
+
+      @Key public String crv;
+
+      @Key public String kid;
+
+      @Key public String kty;
+
+      @Key public String use;
+
+      @Key public String x;
+
+      @Key public String y;
+
+      @Key public String e;
+
+      @Key public String n;
+    }
+
+    PublicKeyLoader(HttpTransportFactory httpTransportFactory) {
+      super();
+      this.httpTransportFactory = httpTransportFactory;
+    }
+
+    @Override
+    public Map<String, PublicKey> load(String certificateUrl) throws Exception {
+      HttpTransport httpTransport = httpTransportFactory.create();
+      JsonWebKeySet jwks;
+      try {
+        HttpRequest request =
+            httpTransport
+                .createRequestFactory()
+                .buildGetRequest(new GenericUrl(certificateUrl))
+                .setParser(GsonFactory.getDefaultInstance().createJsonObjectParser());
+        HttpResponse response = request.execute();
+        jwks = response.parseAs(JsonWebKeySet.class);
+      } catch (IOException io) {
+        return ImmutableMap.of();
+      }
+
+      ImmutableMap.Builder<String, PublicKey> keyCacheBuilder = new ImmutableMap.Builder<>();
+      if (jwks.keys == null) {
+        // Fall back to x509 formatted specification
+        for (String keyId : jwks.keySet()) {
+          String publicKeyPem = (String) jwks.get(keyId);
+          keyCacheBuilder.put(keyId, buildPublicKey(publicKeyPem));
+        }
+      } else {
+        for (JsonWebKey key : jwks.keys) {
+          try {
+            keyCacheBuilder.put(key.kid, buildPublicKey(key));
+          } catch (NoSuchAlgorithmException
+              | InvalidKeySpecException
+              | InvalidParameterSpecException ignored) {
+            ignored.printStackTrace();
+          }
+        }
+      }
+
+      return keyCacheBuilder.build();
+    }
+
+    private PublicKey buildPublicKey(JsonWebKey key)
+        throws NoSuchAlgorithmException, InvalidParameterSpecException, InvalidKeySpecException {
+      if ("ES256".equals(key.alg)) {
+        return buildEs256PublicKey(key);
+      } else if ("RS256".equals((key.alg))) {
+        return buildRs256PublicKey(key);
+      } else {
+        return null;
+      }
+    }
+
+    private PublicKey buildPublicKey(String publicPem)
+        throws CertificateException, UnsupportedEncodingException {
+      return CertificateFactory.getInstance("X.509")
+          .generateCertificate(new ByteArrayInputStream(publicPem.getBytes("UTF-8")))
+          .getPublicKey();
+    }
+
+    private PublicKey buildRs256PublicKey(JsonWebKey key)
```

Comment:
BuildRs256PublicKey and buildEs256PublicKey are the same as the reference implementation. https://github.com/googleapis/google-auth-library-java/blob/1434839e704621dc21f902c81cb67e66f7ef1a05/oauth2_http/java/com/google/auth/oauth2/TokenVerifier.java#L369 Good.